dev_scripts/qa.py

```diff
@@ -0,0 +1,448 @@
+    "7. Dangerzone CLI succeeds in converting multiple documents",
+    "8. Dangerzone can open a document for conversion via right-click -> 'Open With'",
+    "9. Updating Dangerzone handles external state correctly",
+]
+
+
+class Platform(enum.Enum):
+    """List of available platforms for QA."""
+
+    WINDOWS = "windows"
+    UBUNTU_22_04 = "ubuntu-22.04"
+    UBUNTU_22_10 = "ubuntu-22.10"
+    FEDORA_36 = "fedora-36"
+    FEDORA_37 = "fedora-37"
+
+
+class QABase:
+    """Base class for the QA tasks."""
+
+    def __init__(self, try_auto=False, skip_manual=False, debug=False):
+        self.debug = debug
+        self.try_auto = try_auto
+        self.skip_manual = skip_manual
+        self.src = (
+            subprocess.run(
+                [
+                    "git",
+                    "rev-parse",
+                    "--show-toplevel",
+                ],
+                stdout=subprocess.PIPE,
+            )
+            .stdout.decode()
+            .strip("\n")
+        )
+
+    def auto(*msgs, ref=None):
+        """Decorator for running a task automatically.
+
+        This decorator does the following:
+        * Check if the user has asked to run the tasks automatically.
+          - If not, ask the user if they want to try it out.
+          - If yes, run the task (decorated function).
+        * If an exception occurs, allow the user to skip the task, or stop the
+          execution.
+
+        Args:
+            msgs: A list of messages that we should show to the user, one message per
+              line.
+            refs: A link that can be used as a reference for a specific task.
+        """
+
+        def decorator(func):
+            def inner(self, *args, **kwargs):
+                # Assume that the first line is the description of the task, and
+                # decorate it with some "=" signs.
+                dec_msgs = list(msgs)
+                num_equals = (80 - 2 - len(msgs[0])) // 2 or 3
+                dec_msgs.insert(0, "")
+                dec_msgs[1] = "=" * num_equals + " " + msgs[0] + " " + "=" * num_equals
+                self.describe(*dec_msgs)
+
+                # Detect if the task should run automatically, either because the user
+                # has provided the `--try-auto` flag, or because they typed "auto" when
+                # prompted.
+                try_auto = False
+                if self.try_auto:
+                    try_auto = True
+                else:
+                    prompt = "Type 'auto' to try automatically, or ENTER to skip: "
+                    choice = self.prompt(ref=ref, prompt=prompt)
+                    try_auto = "auto" == choice
+
+                # Run the task automatically, if the user asked to, else assume that the
+                # user has run it manually.
+                if try_auto:
+                    logger.info("Trying automatically...")
+                    out = func(self, *args, **kwargs)
```

Comment:
I used to have this code, but it was it felt a bit tricky to get it right:

* The traceback would appear before a prompt.
* Sometimes you wanted to run a command, meaning that you needed to start a new shell.
* It would have to fail if `--try-auto` was used.

So, I figured that we can get this right in a subsequent iteration, as it's not critical. Are you ok with that?

Reply:
Sure!